pkg/bloombuild/planner/planner.go

```diff
@@ -171,39 +183,106 @@ func (p *Planner) runOne(ctx context.Context) error {
 
-		gaps, err := p.findGapsForBounds(ctx, w.tenant, w.table, w.ownershipRange)
-		if err != nil {
-			level.Error(logger).Log("msg", "error finding gaps", "err", err)
-			continue
-		}
+	for table, tenants := range work {
+		for tenant, ownershipRanges := range tenants {
+			logger := log.With(p.logger, "tenant", tenant, "table", table.Addr())
+			tt := tableTenant{
+				tenant: tenant,
+				table:  table,
+			}
+			var tasks []*protos.Task
+
+			// Fetch source metas to be used in both build and cleanup of out-of-date metas+blooms
+			metas, err := p.bloomStore.FetchMetas(
+				ctx,
+				bloomshipper.MetaSearchParams{
+					TenantID: tenant,
+					Interval: bloomshipper.NewInterval(table.Bounds()),
+					Keyspace: v1.NewBounds(0, math.MaxUint64),
+				},
+			)
+			if err != nil {
+				level.Error(logger).Log("msg", "failed to get metas", "err", err)
+				continue
+			}
 
-		now := time.Now()
-		for _, gap := range gaps {
-			totalTasks++
+			for _, ownershipRange := range ownershipRanges {
+				logger := log.With(logger, "ownership", ownershipRange.String())
 
-			task := NewTask(
-				ctx, now,
-				protos.NewTask(w.table, w.tenant, w.ownershipRange, gap.tsdb, gap.gaps),
-			)
+				// Filter only the metas that overlap in the ownership range
+				metasInBounds := bloomshipper.FilterMetasOverlappingBounds(metas, ownershipRange)
+				level.Debug(logger).Log("msg", "found relevant metas", "metas", len(metasInBounds))
 
-			if err := p.enqueueTask(task); err != nil {
-				level.Error(logger).Log("msg", "error enqueuing task", "err", err)
-				continue
+				// Find gaps in the TSDBs for this tenant/table
+				gaps, err := p.findOutdatedGaps(ctx, tenant, table, ownershipRange, metasInBounds, logger)
+				if err != nil {
+					level.Error(logger).Log("msg", "failed to find outdated gaps", "err", err)
+					continue
+				}
+
+				for _, gap := range gaps {
+					tasks = append(tasks, protos.NewTask(table, tenant, ownershipRange, gap.tsdb, gap.gaps))
+				}
+			}
+
+			var enqueuedTasks int
+			resultsCh := make(chan *protos.TaskResult, len(tasks))
+
+			now := time.Now()
+			for _, task := range tasks {
+				queueTask := NewTask(ctx, now, task, resultsCh)
+
+				if err := p.enqueueTask(queueTask); err != nil {
+					level.Error(logger).Log("msg", "error enqueuing task", "err", err)
+					continue
+				}
+
+				totalTasks++
+				enqueuedTasks++
+			}
+
+			tasksResultForTableTenant[tt] = tenantTableTaskResults{
+				totalTasks:    enqueuedTasks,
+				originalMetas: metas,
+				resultsCh:     resultsCh,
 			}
+
+			level.Debug(logger).Log("msg", "enqueued tasks", "tasks", enqueuedTasks)
 		}
 	}
 
 	level.Debug(p.logger).Log("msg", "planning completed", "tasks", totalTasks)
 
+	// Create a goroutine to process the results for each table tenant tuple
+	var wg sync.WaitGroup
+	for tt, results := range tasksResultForTableTenant {
+		wg.Add(1)
+		go func(table config.DayTable, tenant string, results tenantTableTaskResults) {
```

Comment:
This may spawn quite a lot of goroutines.

Idea: Could we use a single results channel? I guess then the problem would be to know when a tenantTable is done, right?

Reply:
> I guess then the problem would be to know when a tenantTable is done, right?
Correct. 

> This may spawn quite a lot of goroutines.
Indeed. On the other hand, I have [this issue][1] open for adding parallel workers to the planner. We can parallelize tenant-tables tuples across those workers and therefore limit the number of goroutines that would be spawned to wait for responses to the number of workers.

So I'd address this there. I'll add a TODO.

[1]: https://github.com/grafana/loki-private/issues/985